app/kuma-dp/pkg/dataplane/dnsproxy/component.go

```diff
@@ -0,0 +1,214 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/miekg/dns"
+
+	"github.com/kumahq/kuma/pkg/core"
+	"github.com/kumahq/kuma/pkg/core/runtime/component"
+	"github.com/kumahq/kuma/pkg/dns/dpapi"
+)
+
+var log = core.Log.WithName("dnsproxy")
+
+type Server struct {
+	// HostPort or unix domain socket for tests
+	address          string
+	done             chan struct{}
+	dnsMap           atomic.Pointer[dnsMap]
+	metrics          *metrics
+	upstreamHostPort string
+	// upstreamHandler is used for testing purposes
+	upstreamHandler func(msg *dns.Msg) (*dns.Msg, error)
+	ready           chan struct{}
+}
+
+func NewServer(address string) *Server {
+	s := Server{
+		address: address,
+		done:    make(chan struct{}),
+		dnsMap:  atomic.Pointer[dnsMap]{},
+		ready:   make(chan struct{}),
+		metrics: newMetrics(),
+	}
+	s.dnsMap.Store(&dnsMap{ARecords: make(map[string]*dnsEntry), AAAARecords: make(map[string]*dnsEntry)})
+	return &s
+}
+
+type dnsMap struct {
+	ARecords    map[string]*dnsEntry
+	AAAARecords map[string]*dnsEntry
+}
+
+type dnsEntry struct {
+	RCode uint16
+	RR    []dns.RR
+}
+
+var _ component.GracefulComponent = &Server{}
+
+func (s *Server) MockDNS(h func(msg *dns.Msg) (*dns.Msg, error)) {
+	s.upstreamHandler = h
+}
+
+func (s *Server) Handler(res dns.ResponseWriter, req *dns.Msg) {
+	start := time.Now()
+	defer func() {
+		s.metrics.RequestDuration.Observe(time.Since(start).Seconds())
+		if err := recover(); err != nil {
+			log.Error(fmt.Errorf("handler panic %v", err), "panic in DNS handler")
+			response := new(dns.Msg)
+			response.SetRcode(req, dns.RcodeServerFailure)
+			_ = res.WriteMsg(response)
+		}
+	}()
+	var response *dns.Msg
+	var dnsEntry *dnsEntry
+	if len(req.Question) > 0 { // Apparently most DNS don't support multiple questions so let's just support the first one
+		if len(req.Question) > 1 {
+			log.Info("[WARNING] multiple questions in a single request, this is not supported", "questions", req.Question)
+		}
+		switch req.Question[0].Qtype {
+		case dns.TypeA:
+			// lookup in our DNS map
+			dnsMap := s.dnsMap.Load()
+			dnsEntry = dnsMap.ARecords[req.Question[0].Name]
+		case dns.TypeAAAA:
+			dnsMap := s.dnsMap.Load()
+			dnsEntry = dnsMap.AAAARecords[req.Question[0].Name]
+		}
+		log.V(1).Info("Got request", "type", req.Question[0].Qtype, "name", req.Question[0].Name, "entry", dnsEntry)
+	}
+	if dnsEntry != nil {
+		response = new(dns.Msg)
+		response.SetRcode(req, int(dnsEntry.RCode))
+		response.Authoritative = true
+		response.Answer = append(response.Answer, dnsEntry.RR...)
+	} else {
+		proxyStart := time.Now()
+		var resp *dns.Msg
+		var err error
+		if s.upstreamHandler != nil {
+			resp, err = s.upstreamHandler(req)
+		} else {
+			c := new(dns.Client)
+			resp, _, err = c.Exchange(req, s.upstreamHostPort)
+		}
+		if err != nil {
+			s.metrics.UpstreamRequestFailureCount.Inc()
+			log.Error(err, "Failed to write message to upstream")
```

Comment:
logs are lowercased https://github.com/kumahq/kuma/blob/5e12d423e9a6f3f6f9b5eefaed4a8759395c14b7/docs/madr/decisions/003-logging.md#L34

Reply:
there are other instances of upper case logs in this file

Reply:
I think I went over all the cases